src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -646,14 +646,22 @@ public void AddDynamicSourceFile(string dynamicFilePath, ImmutableArray<string>
 
                 fileInfo = FixUpDynamicFileInfo(fileInfo, dynamicFilePath);
 
-                // remember map between original dynamic file path to DynamicFileInfo.FilePath
-                _dynamicFilePathMaps.Add(dynamicFilePath, fileInfo.FilePath);
-                _sourceFiles.AddDynamicFile(provider.Value, fileInfo, folders);
-                return;
+                lock (_gate)
+                {
+
+                    // remember map between original dynamic file path to DynamicFileInfo.FilePath
+                    _dynamicFilePathMaps.Add(dynamicFilePath, fileInfo.FilePath);
+                    _sourceFiles.AddDynamicFile(provider.Value, fileInfo, folders);
+                    return;
+                }
             }
+
+            // We have a dynamic file, but no real source file for it to map to, we'll just remember it so later
+            // removals can still work smoothly
+            _dynamicFilePathMaps.Add(dynamicFilePath, null);
```

Comment:
why is this not locked?

Reply:
D'oh.

Reply:
"We have a dynamic file, but no real source file for it to map to"

What does that mean?  When could that happen?  Isn't that bad?  Should we just throw in that case?

Reply:
This code is somewhat akin to source generators; the project system tells us that .cshtml file is there, and we then go and ask the razor file to give us a generated file that matches. If Razor can't provide a file or is unable to or knows that the file shouldn't have generated output, we shouldn't fail. The existing code apparently would have failed later when the project system told us the .cshtml file isn't there anymore.

The separation of concerns means Razor's decision shouldn't impact the project system, but here it is.

Reply:
I'll revise the comment.